File: Medium/486. Predict the Winner.java
class Solution {
    public boolean PredictTheWinner(int[] nums) {
        return Simulator(nums, 0, 0, nums.length - 1, true) >= 0;
    }

    private int Simulator(int[] nums, int score, int left, int right, boolean isMyTurn) {
        if (left == right) {
            return isMyTurn ? score + nums[left] : score - nums[left];
        } else {
            if (isMyTurn) {
                int pickLeft = Simulator(nums, score + nums[left], left + 1, right, false);
                int pickRight = Simulator(nums, score + nums[right], left, right - 1, false);
                return Math.max(pickLeft, pickRight);
            } else {
                int pickLeft = Simulator(nums, score - nums[left], left + 1, right, true);
                int pickRight = Simulator(nums, score - nums[right], left, right - 1, true);
                return Math.min(pickLeft, pickRight);
            }
        }
    }
}

File: Medium/486. Predict the Winner.py
class Solution:
    def predictTheWinner(self, A: List[int]) -> bool:
        @cache
        def dfs(i: int, j: int) -> int:
            if i == j:
                return A[i]
            return max(A[i] - dfs(i + 1, j),
                       A[j] - dfs(i, j - 1))
        return dfs(0, len(A) - 1) >= 0